File: Palace/Views/TPPReloadView.m
#import "UIView+TPPViewAdditions.h"

#import "TPPReloadView.h"
#import "Palace-Swift.h"

@interface TPPReloadView ()

@property (nonatomic) UILabel *messageLabel;
@property (nonatomic) TPPRoundedButton *reloadButton;
@property (nonatomic) UILabel *titleLabel;

@end

static CGFloat const width = 280;

@implementation TPPReloadView

#pragma mark NSObject

- (instancetype)init
{
  self = [super initWithFrame:CGRectMake(0, 0, width, 0)];
  if(!self) return nil;
  
  self.titleLabel = [[UILabel alloc] init];
  self.titleLabel.font = [UIFont boldPalaceFontOfSize:17];
  self.titleLabel.text = NSLocalizedString(@"Connection Failed", nil);
  self.titleLabel.textColor = [UIColor grayColor];
  [self addSubview:self.titleLabel];
  
  self.messageLabel = [[UILabel alloc] init];
  self.messageLabel.numberOfLines = 3;
  self.messageLabel.textAlignment = NSTextAlignmentCenter;
  self.messageLabel.font = [UIFont palaceFontOfSize:12];
  [self setDefaultMessage];
  self.messageLabel.textColor = [UIColor grayColor];
  [self addSubview:self.messageLabel];
  
  self.reloadButton = [[TPPRoundedButton alloc] initWithType:TPPRoundedButtonTypeNormal isFromDetailView:NO];
  [self.reloadButton setTitle:NSLocalizedString(@"Try Again", nil)
                     forState:UIControlStateNormal];
  [self.reloadButton addTarget:self
                        action:@selector(didSelectReload)
              forControlEvents:UIControlEventTouchUpInside];
  [self addSubview:self.reloadButton];
  
  [self layoutIfNeeded];
  
  self.frame = CGRectMake(0, 0, width, CGRectGetMaxY(self.reloadButton.frame));
  
  return self;
}

#pragma mark UIView

- (void)layoutSubviews
{
  [super layoutSubviews];
  CGFloat const padding = 5.0;
  
  {
    [self.titleLabel sizeToFit];
    [self.titleLabel centerInSuperview];
    CGRect frame = self.titleLabel.frame;
    frame.origin.y = 0;
    self.titleLabel.frame = frame;
  }
  
  {
    CGFloat h = [self.messageLabel sizeThatFits:
                 CGSizeMake(CGRectGetWidth(self.frame), CGFLOAT_MAX)].height;
    
    self.messageLabel.frame = CGRectMake(0,
                                         CGRectGetMaxY(self.titleLabel.frame) + padding,
                                         CGRectGetWidth(self.frame),
                                         h);
  }
  
  {
    [self.reloadButton sizeToFit];
    [self.reloadButton centerInSuperview];
    CGRect frame = self.reloadButton.frame;
    frame.origin.y = CGRectGetMaxY(self.messageLabel.frame) + padding;
    self.reloadButton.frame = frame;
  }
}

#pragma mark -

- (void)setDefaultMessage
{
  self.messageLabel.text = NSLocalizedString(@"Check Connection", nil);
  [self setNeedsLayout];
}

- (void)setMessage:(NSString *)msg
{
  self.messageLabel.text = msg;
  [self setNeedsLayout];
}

- (void)didSelectReload
{
  if(self.handler) {
    self.handler();
  }

  [self setDefaultMessage];
}

@end

File: Palace/Settings/DeveloperSettings/TPPReloadView.swift
import UIKit

@objcMembers
final class TPPReloadView: UIView {
  var handler: (() -> Void)?

  // MARK: - UI

  private let titleLabel: UILabel = {
    let label = UILabel()
    label.font = UIFont.boldPalaceFont(ofSize: 17)
    label.text = NSLocalizedString("Connection Failed", comment: "")
    label.textColor = .gray
    return label
  }()

  private let messageLabel: UILabel = {
    let label = UILabel()
    label.numberOfLines = 3
    label.textAlignment = .center
    label.font = UIFont.palaceFont(ofSize: 12)
    label.textColor = .gray
    return label
  }()

  private let reloadButton: TPPRoundedButton = {
    let button = TPPRoundedButton(type: .normal, isFromDetailView: false)
    button.setTitle(NSLocalizedString("Try Again", comment: ""), for: .normal)
    return button
  }()

  // MARK: - Init

  override init(frame _: CGRect) {
    super.init(frame: CGRect(x: 0, y: 0, width: 280, height: 0))
    commonInit()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    commonInit()
  }

  private func commonInit() {
    addSubview(titleLabel)
    addSubview(messageLabel)
    addSubview(reloadButton)

    reloadButton.addTarget(self, action: #selector(didTapReload), for: .touchUpInside)
    setDefaultMessage()
    layoutIfNeeded()
    frame = CGRect(x: 0, y: 0, width: 280, height: reloadButton.frame.maxY)
  }

  // MARK: - Layout

  override func layoutSubviews() {
    super.layoutSubviews()

    let padding: CGFloat = 5
    let width = bounds.width

    // Title
    titleLabel.sizeToFit()
    titleLabel.centerInSuperview()
    var f = titleLabel.frame
    f.origin.y = 0
    titleLabel.frame = f

    // Message
    let messageHeight = messageLabel.sizeThatFits(CGSize(width: width, height: CGFloat.greatestFiniteMagnitude)).height
    messageLabel.frame = CGRect(x: 0, y: titleLabel.frame.maxY + padding, width: width, height: messageHeight)

    // Button
    reloadButton.sizeToFit()
    reloadButton.centerInSuperview()
    var bf = reloadButton.frame
    bf.origin.y = messageLabel.frame.maxY + padding
    reloadButton.frame = bf
  }

  // MARK: - Actions

  @objc private func didTapReload() {
    handler?()
    setDefaultMessage()
  }

  // MARK: - Public (ObjC)

  func setDefaultMessage() {
    messageLabel.text = NSLocalizedString("Check Connection", comment: "")
    setNeedsLayout()
  }

  func setMessage(_ msg: String?) {
    messageLabel.text = msg
    setNeedsLayout()
  }
}
